### DISCUSSION TITLE
[hardhat-zksync-node] Error when deploying contracts to the local node with `hardhat-deploy§

### DISCUSSION THREAD

**Discussion by mmv08** [zkSync-Community-Hub/zksync-developers#482] [SDKs]
### Environment

Testnet

### Acknowledgement

- [X] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

We’re looking to support zksync natively in our hardhat development environment (compilation, testing, deployment). However, we’re blocked by an error that happened during the deployment when trying to run tests on a local in-memory node through `hardhat-zksync-node`:
```
could not coalesce error (error={ "code": -32603, "message": "Internal error" }, payload={ "id": 19, "jsonrpc": "2.0", "method": "eth_sendTransaction", "params": [ { "from": "0x8002cd98cfb563492a6fb3e7c8243b7b9ad4cc92", "gas": "0x3a477c", "nonce": "0x0", "to": "0xe1cb04a0fa36ddd16a06ea828007e35e1a3cbc37", "type": "0x0", "value": "0x5469bd64870200" } ] }, code=UNKNOWN_ERROR, version=6.12.0) {"code":"UNKNOWN_ERROR","error":{"code":-32603,"message":"Internal error"},"payload":{"method":"eth_sendTransaction","params":[{"gas":"0x3a477c","type":"0x0","nonce":"0x0","value":"0x5469bd64870200","from":"0x8002cd98cfb563492a6fb3e7c8243b7b9ad4cc92","to":"0xe1cb04a0fa36ddd16a06ea828007e35e1a3cbc37"}],"id":19,"jsonrpc":"2.0"},"shortMessage":"could not coalesce error"} Error: could not coalesce error (error={ "code": -32603, "message": "Internal error" }, payload={ "id": 19, "jsonrpc": "2.0", "method": "eth_sendTransaction", "params": [ { "from": "0x8002cd98cfb563492a6fb3e7c8243b7b9ad4cc92", "gas": "0x3a477c", "nonce": "0x0", "to": "0xe1cb04a0fa36ddd16a06ea828007e35e1a3cbc37", "type": "0x0", "value": "0x5469bd64870200" } ] }, code=UNKNOWN_ERROR, version=6.12.0)
```

We use hardhat-deploy for deployments, and this seems to be the transaction that funds the deployer address, which carries the deterministic factory deployment

The error can be easily reproduced here: https://github.com/safe-global/safe-smart-account/pull/742
by running
```
npm i && npm run test:zk
```

Nothing helpful in the `era_test_node.log` either.

### Expected Behavior

1. The contracts are deployed and the tests are run, similar to how it works with a regular hardhat node.
2. If something is truly wrong with the configuration, a meaningful error message is shown 

### Code Example

Code example is available here:
https://github.com/safe-global/safe-smart-account/pull/742

### Repo Link (Optional)

https://github.com/safe-global/safe-smart-account/pull/742

**Comment by bxpana** [comment]
Thanks for the post. We'll take a look and get back to you guys.

**Reply by mmv08** [reply]
We came up with a more minimal reproduction example, seems like the node is not supporting `eth_sendTransaction`.

```
import hre from "hardhat"

export const getWallet = async (privateKey?: string) => {
  const wallet = await hre.zksyncEthers.getWallet(privateKey ?? process.env.WALLET_PRIVATE_KEY)

  return wallet
}

describe("Transaction", function () {
  it.only("Should send a transaction", async function () {
    const wallet = await getWallet(
      "0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110"
    )

    const tx = await wallet.populateTransaction({
      to: `0x${"0".repeat(39)}1`,
      data: "0x",
      value: 1,
    })
    tx.value = "0x" + tx.value?.toString(16)
    tx.gasLimit = "0x" + tx.gasLimit?.toString(16)
    tx.gasPrice = "0x" + tx.gasPrice?.toString(16)
    tx.nonce = "0x" + tx.nonce?.toString(16)
    tx.chainId = "0x" + tx.chainId?.toString(16)
    tx.type = "0x0"

    const receipt = await hre.network.provider.send("eth_sendTransaction", [tx])
    console.log({ receipt })
  })
})
```
fails with 
```
  1) Transaction
       Should send a transaction:
     Error: could not coalesce error (error={ "code": -32603, "message": "Internal error" }, payload={ "id": 2, "jsonrpc": "2.0", "method": "eth_sendTransaction", "params": [ { "chainId": "0x104", "data": "0x", "from": "0x36615Cf349d7F6344891B1e7CA7C72883F5dc049", "gasLimit": "0x359b19", "gasPrice": "0x2faf080", "nonce": "0x0", "to": "0x0000000000000000000000000000000000000001", "type": "0x0", "value": "0x1" } ] }, code=UNKNOWN_ERROR, version=6.12.0)
      at makeError (node_modules/ethers/src.ts/utils/errors.ts:694:21)
      at Provider.getRpcError (node_modules/ethers/src.ts/providers/provider-jsonrpc.ts:1049:25)
      at /Users/mmv/projects/safe/test-greeter/node_modules/ethers/src.ts/providers/provider-jsonrpc.ts:563:45
      at processTicksAndRejections (node:internal/process/task_queues:95:5)
```

**Reply by bxpana** [reply]
Thank you, will relay that to the team as well.

**Comment by bxpana** [comment]
Turns out `eth_sendTransaction` is **not** supported on testnet and mainnet. It is supported for era-test-node to change the owner address to impersonate accounts for testing purposes.

**Reply by mmv08** [reply]
Does that mean that the node is not suitable for hardhat testing? It would've been nice if it would've been plug n play

**Reply by bxpana** [reply]
The in-memory node (era-test-node) should support `eth_sendTransaction` for testing. It's only testnet and mainnet that do not support it.

**Reply by mmv08** [reply]
But in the reproduction example, we use the test node. Should it be supported? Or did we misconfigure something?

**Reply by bxpana** [reply]
Apologies for the delay here. Right now everything that you can test with the in-memory node is documented here https://github.com/matter-labs/era-test-node/blob/main/SUPPORTED_APIS.md

Let us know if that helps.

**Reply by mmv08** [reply]
Well, here it says that `eth_sendTransaction` is supported, but in our case it throws an error (which could be found in the original post)

**Reply by MexicanAce** [reply]
I believe a manually populated transaction will fail with the ZKEVM, you will need to update your transaction with the appropriate signer, example:

```typescript
    await provider.send("hardhat_impersonateAccount", [RichAccount]);

    const signer = await ethers.getSigner(RichAccount);
    const tx = {
      to: userWallet.address,
      value: ethers.utils.parseEther("0.3"),
    };

    const recieptTx = await signer.sendTransaction(tx);
    await recieptTx.wait();
```

**Reply by mmv08** [reply]
The `eth_sendTransaction` in question is handled by the `hardhat-deploy` package, we cannot alter the code of the package to use another signer.

**Reply by MexicanAce** [reply]
@mmv08 I would try adding the line
```ts
await provider.send("hardhat_impersonateAccount", ["0x7726827caac94a7f9e1b160f7ea819f172f7b6f9d2a97f992c38edeab82d4110"]);
```
before you try to call 
```ts
const receipt = await hre.network.provider.send("eth_sendTransaction", [tx])
```

**Reply by mmv08** [reply]
@MexicanAce this is not possible, since the deployment is handled by the hardhat-deploy package. The same code works fine with the dockerized node and any other network, therefore I believe there is a bug

**Reply by bxpana** [reply]
> Well, here it says that `eth_sendTransaction` is supported, but in our case it throws an error (which could be found in the original post)

To my understanding it is supported but you have to directly call the API endpoint not just using a library that uses the endpoint.

**Reply by mmv08** [reply]
> > Well, here it says that `eth_sendTransaction` is supported, but in our case it throws an error (which could be found in the original post)
> 
> To my understanding it is supported but you have to directly call the API endpoint not just using a library that uses the endpoint.

Guys, I'm sorry, but this is ridiculous. You can find my reproduction [example](https://github.com/zkSync-Community-Hub/zksync-developers/discussions/482#discussioncomment-9223323) above where I called the endpoint directly and it didn't work.

**Reply by bxpana** [reply]
Took some time to dive into this and got things working. Used your reproduction example and put the `hardhat_impersonateAccount` before calling `eth_sendTransaction` and was able to successfully send the transaction. You guys can test it out here https://github.com/bxpana/safe-smart-account/blob/v1.5.0-zksync/test/zktest.ts

**Reply by mmv08** [reply]
What is the point of this limitation @bxpana? I'm using a private key of a special account provided by the [node](https://github.com/matter-labs/era-test-node?tab=readme-ov-file#-using-rich-wallets)

The tutorial doesn't mention this step either: https://docs.zksync.io/build/test-and-debug/hardhat#run-tests-with-hardhat

Why is the additional step needed if the package claims compatibility with hardhat and my mnemonic/pk is already provided in the hardhat configuration? I strongly believe this behaviour is suboptimal, hurting developer experience, and blocking integrations for zksync project (as I mentioned, we cannot alter the library code to execute `hardhat_impersonateAccount`, neither should we, in my opinion, since it's redundant)

**Reply by mmv08** [reply]
I agree it might be outside of the node responsibility, but IMO the [hardhat-zksync-node](https://github.com/matter-labs/hardhat-zksync/tree/main/packages/hardhat-zksync-node) plugin should handle all the necessary setup steps

**Reply by mmv08** [reply]
Also in the docs it's clearly stated:
> hardhat-deploy version ^0.11.26 supports deployments on ZKsync Era.

https://docs.zksync.io/build/tooling/hardhat/migrating-to-zksync#deploy-contracts

**Reply by bxpana** [reply]
> What is the point of this limitation @bxpana? I'm using a private key of a special account provided by the [node](https://github.com/matter-labs/era-test-node?tab=readme-ov-file#-using-rich-wallets)

I'm not sure about the limitation here and will go back to the team to see if we can get some more resources here to help you guys out.

**Reply by bxpana** [reply]
I believe we have a chat going with you guys so I'm going to close this one out for now.
